From 3f1af57eaefadc5a28bc7feede9721b0cecdc101 Mon Sep 17 00:00:00 2001
From: Neil Horman <nhorman@redhat.com>
Date: Wed, 22 Oct 2014 18:03:06 -0400
Subject: [net] usb/huawei_cdc_ncm: increase command buffer size

Message-id: <1414000989-9522-3-git-send-email-nhorman@redhat.com>
Patchwork-id: 98072
O-Subject: [RHEL7.1 PATCH v2 2/5] net: huawei_cdc_ncm: increase command buffer size
Bugzilla: 1129796
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>

Messages from the modem exceeding 256 bytes cause communication
failure.

The WDM protocol is strictly "read on demand", meaning that we only
poll for unread data after receiving a notification from the modem.
Since we have no way to know how much data the modem has to send,
we must make sure that the buffer we provide is "big enough".
Message truncation does not work. Truncated messages are left unread
until the modem has another message to send.  Which often won't
happen until the userspace application has given up waiting for the
final part of the last message, and therefore sends another command.

With a proper CDC WDM function there is a descriptor telling us
which buffer size the modem uses. But with this vendor specific
implementation there is no known way to calculate the exact "big
enough" number.  It is an unknown property of the modem firmware.
Experience has shown that 256 is too small.  The discussion of
this failure ended up concluding that 512 might be too small as
well. So 1024 seems like a reasonable value for now.

Fixes: 41c47d8cfd68 ("net: huawei_cdc_ncm: Introduce the huawei_cdc_ncm driver")
Cc: Enrico Mioso <mrkiko.rs@gmail.com>
Reported-by: Dan Williams <dcbw@redhat.com>
Signed-off-by: Bjorn Mork <bjorn@mork.no>
Acked-By: Enrico Mioso <mrkiko.rs@gmail.com>
Tested-by: Dan Williams <dcbw@redhat.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 3acc74619b0175b7a154cf8dc54813f6faf97aa9)

BZ: 1129796
Brew: http://brewweb.devel.redhat.com/brew/taskinfo?taskID=8150596
Testing: Tested by reporter with success
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/usb/huawei_cdc_ncm.c b/drivers/net/usb/huawei_cdc_ncm.c
index 312178d..75d9ee4 100644
--- a/drivers/net/usb/huawei_cdc_ncm.c
+++ b/drivers/net/usb/huawei_cdc_ncm.c
@@ -84,12 +84,13 @@ static int huawei_cdc_ncm_bind(struct usbnet *usbnet_dev,
  ctx = drvstate->ctx;
 
  if (usbnet_dev->status)
-  /* CDC-WMC r1.1 requires wMaxCommand to be "at least 256
-   * decimal (0x100)"
+  /* The wMaxCommand buffer must be big enough to hold
+   * any message from the modem. Experience has shown
+   * that some replies are more than 256 bytes long
    */
   subdriver = usb_cdc_wdm_register(ctx->control,
        &usbnet_dev->status->desc,
-       256, /* wMaxCommand */
+       1024, /* wMaxCommand */
        huawei_cdc_ncm_wdm_manage_power);
  if (IS_ERR(subdriver)) {
   ret = PTR_ERR(subdriver);
-- 
1.7.1